fabtests/prov/efa/src/rdm_rnr_queue_resend.c

```diff
@@ -0,0 +1,438 @@
+		for (i = 0; i < expected_rnr_error; i++) {
+			switch(atomic_op) {
+			case FI_ATOMIC_WRITE:
+				ret = ft_post_atomic(FT_ATOMIC_BASE, ep, NULL, NULL, NULL, NULL,
+			                        &remote, FI_INT32, FI_ATOMIC_WRITE, &fi_ctx_atomic);
+				break;
+			case FI_ATOMIC_READ:
+				ret = ft_post_atomic(FT_ATOMIC_FETCH, ep, NULL, NULL, result,
+						fi_mr_desc(mr_result), &remote, FI_INT32,
+						FI_ATOMIC_READ, &fi_ctx_atomic);
+				break;
+			case FI_CSWAP:
+				ret = ft_post_atomic(FT_ATOMIC_COMPARE, ep,
+						compare, fi_mr_desc(mr_compare),
+						result, fi_mr_desc(mr_result), &remote, FI_INT32,
+						FI_CSWAP, &fi_ctx_atomic);
+				break;
+			default:
+				FT_ERR("Unknown atomic op type\n");
+				return EXIT_FAILURE;
+			}
+			if (ret)
+				return ret;
+		}
+	} else {
+		for (i = 0; i < expected_rnr_error; i++) {
+			ret = ft_post_tx(ep, remote_fi_addr, opts.transfer_size, NO_CQ_DATA, &tx_ctx);
+			if (ret)
+				return ret;
+		}
+	}
+	return ret;
+}
+
+static int rnr_queue_resend_test(int req_pkt, enum fi_op atomic_op)
+{
+	int ret, i;
+	void *result = NULL;
+	void *compare = NULL;
+	struct fid_mr *mr_result = NULL;
+	struct fid_mr *mr_compare = NULL;
+
+	if (atomic_op) {
+		ret = alloc_atomic_res(fi, &result, &compare, &mr_result, &mr_compare);
+		if (ret) {
+			FT_PRINTERR("alloc_ep_res_atomic()", -ret);
+			return ret;
+		}
+	}
+	if (opts.rma_op || atomic_op) {
+		ret = ft_exchange_keys(&remote);
+		if (ret) {
+			FT_PRINTERR("ft_exchange_keys()", -ret);
+			return ret;
+		}
+	} else {
+		ret = ft_sync();
+		if (ret) {
+			FT_PRINTERR("ft_sync()", -ret);
+			return ret;
+		}
+	}
+	/*
+	 * Wait 1s here to ensure the server receives RXR_RECEIPT_PKT if delivery_complete
+	 * is requested, before the client starts sending (fi->rx_attr->size) packets.
+	 * Without this, the server might already receive multiple packets, before
+	 * receiving RXR_RECEIPT_PKT, which causes the server's internally
+	 * pre-posted rx buffer not getting run out.
+	 */
+	sleep(1);
+	if (opts.dst_addr) {
+		/*
+		 * Post fi->rx_attr->size sends on client side to run out of pre-posted
+		 * internal buffer on server side.
+		 */
+		for (i = 0; i < fi->rx_attr->size; i++) {
+			ret = ft_post_tx(ep, remote_fi_addr, 32, NO_CQ_DATA, &tx_ctx);
+			if (ret)
+				return ret;
+		}
+		/*
+		 * For REQ packet types, post send/rma/atomic operation here to trigger RNR
+		 * queueing and re-sending logic happening.
+		 */
+		if (req_pkt) {
+			ret = trigger_rnr_queue_resend(atomic_op, result, compare, mr_result, mr_compare);
+			if (ret)
+				goto out;
+		} else if (!opts.rma_op && !atomic_op) {
+			for (i = 0; i < expected_rnr_error; i++) {
+				ret = ft_rx(ep, opts.transfer_size);
+				if (ret)
+					return ret;
+			}
+		}
+		ret = ft_get_tx_comp(tx_seq);
+		if (ret)
+			return ret;
+	} else {
+		/*
+		 * For non-REQ packet types, post send/rma/atomic operation here so that RNR
+		 * queuing/re-sending logic can be triggered on client side with non-REQ
```

Comment:
The comment is hard to understand. I think what you want to say is: 
"server as the receiver/responder of an two-sided/one-sided operation, will send non-REQ packet back to client. To ensure non-REQ packet will encounter RNR, server post enough TX operations to client to occupy client's pre-posted receive buffer."

Reply:
No. In this case, the server sends a REQ packet to the client, and the client is expected to send back a non-REQ packet (e.g., CTS), which is expected to trigger RNR on the client side.

Reply:
Can you update the comment with the exact process? who send non-req? who will encounter RNR? etc

Reply:
Update the comments for REQ packet and non-REQ packet, respectively, by providing procedure explanation and example.